### ISSUE TITLE
Multiple declarations (for the same variable) located in different code blocks of `if ... else` should not cause 'no-redeclare' errors

### ISSUE THREAD

**Issue by darkred** [eslint/eslint#7340]
**Tell us about your environment**

* **ESLint Version:**   ESlint 3.7.1
* **Node Version:** 6.7.0
* **npm Version:** 3.10.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

On the following code I only have this rule:
```
/*eslint no-redeclare: "error"*/
```

**What did you do? Please include the actual source code causing the issue.**

I have this code:
```
/*eslint no-redeclare: "error"*/
var z;
if (z === true) {
	var a = 1;
} else {
	var a = 2;
}
```

**What did you expect to happen?**

Because `z === true` is not true, its block of code ` var a = 1` is not executed: the 2nd block of code `    var a = 2` is executed instead.
Therefore no `no-redeclare` error should occur for `var a = 2`,
i.e. I believe that multiple declarations of the same variable 
located in different code blocks of `if ... else` should not cause `no-redeclare` errors,
because when running the code, the 1st condition (containing the first declaration) can be false.

**What actually happened? Please include the actual, raw output from ESLint.**

I get this error: 
```
a is already defined (no-redeclare )     Line 6, Column 9
```

**Comment by ilyavolodin** [comment]
JavaScript uses hoisting for variable declarations (unless you are using `let`/`const`). On top of that, JavaScript is function-scoped, so all of the declarations will be moved to the top of the function (not a block). So you `a` variable will be declared twice once the code is re-written with hoisting in mind. The rule is behaving correctly in this case, because that's what it's built to prevent. If you want to set the value of a variable based on the condition, you should move variable declaration outside of the condition. i.e.
```js
var a = z ? 1 : 2;

or

var z, a;
if (z === true) {
  a = 1;
} else {
  a = 2;
}
```

**Comment by darkred** [comment]
Alright, thanks for the information.
